yktracec/src/jitmodbuilder.cc

```diff
@@ -1912,6 +1920,63 @@ class JITModBuilder {
     finalise(AOTMod, &Builder);
     return JITMod;
   }
+
+  /// Handle promoted variables.
+  void handlePromote(CallInst *CI, MappableFrame *MPF, BasicBlock *BB,
+                     size_t CurBBIdx, size_t CurInstrIdx) {
+    // First lookup the constant value the trace is going to use.
+    uint64_t PConst = __yk_lookup_promote_usize();
+    Value *PConstLL =
+        Builder.getIntN(PointerSizedIntTy->getIntegerBitWidth(), PConst);
+    Value *PromoteVar = CI->getArgOperand(0);
+    Value *JITPromoteVar = getMappedValue(PromoteVar);
+
+    // By promoting a value to a constant, we specialised the remainder of the
+    // trace to that constant value. We must emit a guard that ensures that we
+    // doptimise if the promoted value deviates from the "baked-in" constant.
+    MPF->LastSMCall = cast<CallInst>(CI->getNextNonDebugInstruction());
+
+    // When we encounter `%y = yk_promote(%x)` in the AOT module, we don't copy
+    // this instruction into the JIT module, only update future references to
+    // `%y` with a constant. However, in AOT `%y` still exists and needs to be
+    // re-materialised when we deoptimise, so we have to point some JIT variable
+    // back to `%y` in our AOTMAP.
+    //
+    // Since in AOT, `%y` and `%x` are identical, really we'd like to point
+    // `AOTMap[JITPromoteVar]` to `CI`, but that would overwrite the existing
+    // (required!) mapping for `JITPromoteVar`.
+    //
+    // What we therefore do is make an instruction to the effect `%y = %x` in
+    // the JIT module, and map that instruction back to `%y` in the AOT module.
+    Instruction *NewInst = SelectInst::Create(
+        ConstantInt::get(Type::getInt1Ty(JITMod->getContext()), 0),
+        JITPromoteVar, JITPromoteVar);
+    Builder.Insert(NewInst);
+
+    VMap[CI] = NewInst;
+    insertAOTMap(CI, NewInst, CurBBIdx, CurInstrIdx);
+
+    BasicBlock *FailBB = getGuardFailureBlock(BB, CurBBIdx, CI, CurInstrIdx);
+    BasicBlock *SuccBB = BasicBlock::Create(JITMod->getContext(),
+                                            GUARD_SUCCESS_BLOCK_NAME, JITFunc);
+    Value *Deopt = Builder.CreateICmp(CmpInst::Predicate::ICMP_NE,
+                                      JITPromoteVar, PConstLL);
+    Builder.CreateCondBr(Deopt, FailBB, SuccBB);
+
+    // Carry on constructing the trace module in the success block.
+    Builder.SetInsertPoint(SuccBB);
+
+    // Update the VMap so that henceforth, the promoted variable is in
```

Comment:
Are we tracking variables here?

Reply:
Similarly to before, value and variable are (pretty much) the same at the llvm-level.

Reply:
Maybe say "LLVM Value" to disambiguate?

Reply:
Fixed in 0d7133a24584c86f306688cdaa5fa1172ccc2306

Reply:
Is it worth (not in this PR!) homogenising that at some point?

Reply:
I'm not against it.

Reply:
OK.